sdpa_op.py

```diff
@@ -0,0 +1,177 @@
+import torch
+from torch._subclasses import FakeTensorMode
+from torch.fx.experimental.symbolic_shapes import ShapeEnv, DimDynamic
+from torch._functorch.eager_transforms import _unwrap_all_tensors_from_functional, _wrap_all_tensors_to_functional, functionalize
+import torch.nn.functional as F
+import torch.fx as fx
+from torch._ops import HigherOrderOperator
+from torch._C import DispatchKey, DispatchKeySet, _ExcludeDispatchKeyGuard
+import torch.utils._pytree as pytree
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+from torch.utils._python_dispatch import (
+    _get_current_dispatch_mode,
+    _pop_mode_temporarily,
+)
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch._inductor.compile_fx import compile_fx_inner
+
+sdpa = HigherOrderOperator("sdpa")
+
+def math_attention(q, k, v, score_mod, *other_buffers):
+    scores = q @ k.transpose(-2, -1)
+
+    from functorch.dim import dims
+    b, h, m, n = dims()
+
+    scores = scores[b, h, m, n]
+    scores = score_mod(scores, b, h, m, n, *other_buffers)
+    scores = scores.order(b, h, m, n)
+
+    scores = scores.softmax(dim=-1)
+    return scores @ v
+
+@sdpa.py_impl(DispatchKey.CompositeExplicitAutograd)
+def sdpa_dense(q, k, v, score_mod, *other_buffers):
+    return math_attention(q, k, v, score_mod, *other_buffers).contiguous()
+    # out = F.scaled_dot_product_attention(q, k, v, scale=1).contiguous()
+    # return out
+
+@sdpa.py_impl(DispatchKey.Autograd)
+def sdpa_autograd(*args, **kwargs):
+    with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.AutogradCPU)):
```

Comment:
Use _DispatchBelowAutograd: it'll also add the other autograd keys, which are necessary. I'm kind of confused that this PR works even though we've only excluded AutogradCPU.

Is the spda op user-facing? If so, then we want this to error on backward if the inputs do require grad. Example of how to do this over at https://github.com/pytorch/pytorch/blob/dc19f8a6b5d051cd01a1fad6bfadb898ba466f6e/functorch/experimental/_cond.py#L122-L153.

For more brownie points (no need to do it if you're in a rush) we should actually just refactor that code to be a generic "autograd not implemented" function in torch._higher_order_ops.utils that can be shared among all higher order ops.

Reply:
To get more familiar I will work on implementing that